model/storage/src/main/java/org/keycloak/storage/adapter/AbstractUserAdapterFederatedStorage.java

```diff
@@ -61,6 +62,7 @@ public abstract class AbstractUserAdapterFederatedStorage extends UserModelDefau
     protected KeycloakSession session;
     protected RealmModel realm;
     protected ComponentModel storageProviderModel;
+    private MultivaluedHashMap<String, String> federatedAttributes;
```

Comment:
IMHO the attributes shouldn't be cached on the instance of a user adapter, as the constructor for this class can be called multiple multiple times during a session for the same user. It should either be stateless as the JPA's `UserAdapter`, or should ensure that the data it caches is consistent within a `KeycloakSession`.

Reply:
See org.keycloak.models.cache.infinispan.UserCacheSession#cacheUser. This is the place where the attributes are queried multiple times. AFAIK, we won't have adapter instances created multiple times within the `KeycloakSession` because the user will be cached already.

Reply:
Wrappers are created also when you modify a cached user AFAIK. I'd still would vote against it, at least for KC26 forward. It might be a hack for a backport though.